shell/platform/android/io/flutter/embedding/FlutterEngine.java

```diff
@@ -0,0 +1,262 @@
+package io.flutter.embedding;
+
+import android.app.Activity;
+import android.content.Context;
+import android.content.res.Resources;
+import android.util.Log;
+
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import io.flutter.embedding.legacy.FlutterPluginRegistry;
+import io.flutter.plugin.common.BinaryMessenger;
+import io.flutter.view.FlutterRunArguments;
+
+/**
+ * A single Flutter execution environment.
+ *
+ * A {@code FlutterEngine} can execute in the background, or it can be rendered to the screen by
+ * using the accompanying {@link FlutterRenderer}.  Rendering can be started and stopped, thus
+ * allowing a {@code FlutterEngine} to move from UI interaction to data-only processing and then
+ * back to UI interaction.
+ *
+ * To start running Flutter within this {@code FlutterEngine}, use {@link #runFromBundle(FlutterRunArguments)}.
+ * The {@link #runFromBundle(FlutterRunArguments)} method may not be invoked twice on the same
+ * {@code FlutterEngine}.
+ *
+ * To start rendering Flutter content to the screen, use {@link #getRenderer()} to obtain a
+ * {@link FlutterRenderer} and then attach a {@link FlutterRenderer.RenderSurface}.  Consider using
+ * a {@link FlutterView} as a {@link FlutterRenderer.RenderSurface}.
+ * TODO(mattcarroll): for the above instructions for RenderSurface to be true, we need to refactor
+ *                    the native relationships and also the PlatformViewsController relationship.
+ */
+public class FlutterEngine implements BinaryMessenger {
+  private static final String TAG = "FlutterEngine";
+
+  private final Resources resources;
+  private final FlutterJNI flutterJNI;
+  private final FlutterRenderer renderer;
+  private final FlutterPluginRegistry pluginRegistry;
+  private final Map<String, BinaryMessageHandler> mMessageHandlers;
+  private final Map<Integer, BinaryReply> mPendingReplies = new HashMap<>();
+  private long nativeObjectReference;
+  private boolean isBackgroundView; // TODO(mattcarroll): rename to something without "view"
+  private boolean applicationIsRunning;
+  private int mNextReplyId = 1;
+
+  FlutterEngine(
+      Context context,
+      Resources resources,
+      boolean isBackgroundView
+  ) {
+    this.flutterJNI = new FlutterJNI();
+    this.resources = resources;
+    this.isBackgroundView = isBackgroundView;
+    pluginRegistry = new FlutterPluginRegistry(this, context);
+    mMessageHandlers = new HashMap<>();
+
+    attach();
+    // TODO(mattcarroll): FlutterRenderer is temporally coupled to attach(). Remove that coupling if possible.
+    this.renderer = new FlutterRenderer(flutterJNI, nativeObjectReference);
+  }
+
+  private void attach() {
+    // TODO(mattcarroll): what impact does "isBackgroundView' have?
```

Comment:
Yeah, I'm equally unclear.  I think maybe this requires a conversation with @bkonyi

Reply:
`isBackgroundView` determined whether or not a `FlutterNativeView` had a rendering surface specified and changed the number of threads associated with the view based on this.

FYI, `FlutterNativeView` is used directly in plugin code so this is likely to break background execution. We'll need to ensure we don't break the `android_alarm_manager` plugin again, since that breakage went unnoticed for quite awhile and was a major pain point for external devs relying on that functionality.

Reply:
@bkonyi Do you happen to have any ideas as to how we could reasonably get this embedding code under test?  I've been asking everyone I can think of and so far no one really has any ideas.  This code is definitely approaching an intersection with many other responsibilities and it definitely needs tests to prevent regressions.

The problem for me is that I'm used to leveraging test infrastructure that is immediately available via Gradle and Android Studio, but this embedding code is sitting in a non-project-structure, built by .gn files...

Reply:
Honestly, I'm not totally sure what would be the best way to test this. The only method I can think of is creating a test which spawns a background isolate which is then used to send a message over a `SendPort` to the main isolate to confirm that the creation of the background isolate was successful. However, this is more of a sanity check than anything.

Reply:
@bkonyi I've created a sample app that integrates almost all of our plugins.  Most of them work fine.  I haven't tried alarm manager.  I think it might be helpful for us to sit down together sometime this week and test some background execution.  I'm not up to date with the form that takes on Android, but I'm guessing some changes will be necessary to support the new division of responsibilities within the embedding.